kart/diff_util.py

```diff
@@ -1,12 +1,77 @@
+import functools
 import logging
 
+import pygit2
+
 from .diff_structs import DatasetDiff, RepoDiff
 from .key_filters import DatasetKeyFilter, RepoKeyFilter
 from .structure import RepoStructure
 
 L = logging.getLogger("kart.diff_util")
 
 
+class WCDiffContext:
+    """
+    Context shared between datasets during a single diff operation.
+    This could be used for overriding what the working copy is or where it is found,
+    but it is mostly useful for caching the results of diff operations that could be useful to
+    more than a single dataset, but must have a shorter-lived scope than the repository itself.
+
+    The operation of diffing the user's worktree with the worktree index is just such an operation -
+    - we don't want to run it up front, in case there are no datasets that need this info
+    - we want to run it as soon a the first dataset needs this info, then cache the result
+    - we want the result to stay cached for the duration of the diff operation, but no longer
+      (in eg a long-running test, there might be several diffs run and the worktree might change)
+    """
+
+    def __init__(self, repo, all_ds_paths=None):
+        self.repo = repo
+        self.is_bare = repo.is_bare
+        self.all_ds_paths = all_ds_paths
+
+    @property
+    def table_working_copy(self):
+        assert not self.is_bare
+        return self.repo.working_copy
+
+    @property
+    def workdir_path(self):
+        assert not self.is_bare
+        return self.repo.workdir_path
+
+    @property
+    def workdir_index_path(self):
+        return self.repo.gitdir_file("worktree-index")
+
+    @functools.lru_cache(maxsize=1)
+    def workdir_diff(self):
+        # This is the main reason to hold onto this context throughout an entire diff -
+        # we can reuse this result for more than one dataset.
+        index = pygit2.Index(str(self.workdir_index_path))
+        index._repo = self.repo
+        return index.diff_to_workdir(pygit2.GIT_DIFF_INCLUDE_UNTRACKED)
+
+    @functools.lru_cache(maxsize=1)
+    def workdir_deltas_by_ds_path(self):
+        """Returns all the deltas from self.workdir_diff() but grouped by dataset path."""
+        with_and_without_slash = [
+            (p.rstrip("/") + "/", p.rstrip("/")) for p in self.all_ds_paths
+        ]
+
+        def find_ds_path(delta):
```

Comment:
Could this function be reduced to basically `return path.split('/.')[0]`, or more explicitly, `re.split(r'/\.(table|point-cloud)-dataset', path)[0]` ?

Reply:
No - the ODB has helpful marker directories that tell you when you are entering a dataset, which have dots in them meaning "I am hidden".
This code deals with the user-visible file-based working-copy (aka workdir aka worktree, we probably need to standardise a bit). That doesn't have the markers, or any hidden folders - so the way we can tell which files belong in which datasets is by seeing is by getting a list of dataset paths from the ODB side of things, and then comparing delta paths to those dataset paths to see which dataset each one is inside, if any.